tx_applier_test.go

```diff
@@ -0,0 +1,191 @@
+// FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+// COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+// IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+// CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+package wavelet
+
+import (
+	"bytes"
+	"encoding/binary"
+	"github.com/perlin-network/noise/skademlia"
+	"github.com/perlin-network/wavelet/avl"
+	"github.com/perlin-network/wavelet/store"
+	"github.com/perlin-network/wavelet/sys"
+	"github.com/stretchr/testify/assert"
+	"math/rand"
+	"testing"
+)
+
+type TxApplierTestAccount struct {
+	keys   *skademlia.Keypair
+	effect struct {
+		Balance uint64
+		Stake   uint64
+	}
+}
+
+func TestApplyTransaction_Single(t *testing.T) {
+	const InitialBalance = 100000000
+
+	state := avl.New(store.NewInmem())
+	var initialRoot *Transaction
+
+	viewID := uint64(0)
+	state.SetViewID(viewID)
+
+	accounts := make(map[AccountID]*TxApplierTestAccount)
+	accountIDs := make([]AccountID, 0)
+	for i := 0; i < 60; i++ {
+		keys, err := skademlia.NewKeys(1, 1)
+		assert.NoError(t, err)
+		account := &TxApplierTestAccount{
+			keys: keys,
+		}
+		if i == 0 {
+			initialRoot = AttachSenderToTransaction(keys, NewTransaction(keys, sys.TagNop, nil))
+		}
+		WriteAccountBalance(state, keys.PublicKey(), InitialBalance)
+		account.effect.Balance = InitialBalance
+
+		accounts[keys.PublicKey()] = account
+		accountIDs = append(accountIDs, keys.PublicKey())
+	}
+
+	rng := rand.New(rand.NewSource(42))
+
+	round := NewRound(viewID, state.Checksum(), 0, &Transaction{}, initialRoot)
+
+	for i := 0; i < 10000; i++ {
+		switch rng.Intn(2) {
+		case 0:
+			amount := rng.Uint64()%100 + 1
+			account := accounts[accountIDs[rng.Intn(len(accountIDs))]]
+			account.effect.Stake += amount
+			account.effect.Balance -= amount
+
+			var intBuf [8]byte
+			payload := bytes.NewBuffer(nil)
+			payload.WriteByte(sys.PlaceStake)
+			binary.LittleEndian.PutUint64(intBuf[:8], uint64(amount))
+			payload.Write(intBuf[:8])
+
+			tx := AttachSenderToTransaction(account.keys, NewTransaction(account.keys, sys.TagStake, payload.Bytes()))
+			err := ApplyTransaction(&round, state, tx)
+			assert.NoError(t, err)
+		case 1:
+			amount := rng.Uint64()%100 + 1
+			fromAccount := accounts[accountIDs[rng.Intn(len(accountIDs))]]
+			toAccount := accounts[accountIDs[rng.Intn(len(accountIDs))]]
+			if fromAccount.keys.PublicKey() == toAccount.keys.PublicKey() {
+				continue
+			}
+
+			toAccountID := toAccount.keys.PublicKey()
+
+			fromAccount.effect.Balance -= amount
+			toAccount.effect.Balance += amount
+
+			payload := bytes.NewBuffer(nil)
+			payload.Write(toAccountID[:])
+			var intBuf [8]byte
+			binary.LittleEndian.PutUint64(intBuf[:], amount)
+			payload.Write(intBuf[:])
+
+			tx := AttachSenderToTransaction(fromAccount.keys, NewTransaction(fromAccount.keys, sys.TagTransfer, payload.Bytes()))
```

Comment:
I do not see any tests for other transaction types apart from `sys.TagTransfer` transactions.

Reply:
Also sys.TagStake. These tests are only for general transaction applying routines, but not for specific tags.

Reply:
We should test not only for general transaction applying routines. The most important and sensitive logic we should whitebox test for example is gas fee deduction, and recursively calling a smart contract function.

Reply:
These will need calling a full smart contract - better do this in another PR.